mautic/mautic
# Fix timezone segment rebuild

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" for all features, enhancements and bug fixes (until 3.3.0 is released) 
| Bug fix?                               | yes/no
| New feature?                           | yes/no
| Deprecations?                          | yes/no
| BC breaks?                             | yes/no
| Automated tests included?              | yes/no
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 




[//]: # ( As applicable: )
#### Steps to reproduce the bug:
Apply this fix first https://github.com/mautic/mautic/pull/7752
1. Setup local timezone on server to UTC
2. Setup timezone in Mautic configuration or in local.php to Europe/Paris
3. Create two custom field `date `and `datetime `format
4. Then create segment filter with both fields and relative dates
- equal = 2 days
- greater/equal 2 days

![image](https://user-images.githubusercontent.com/462477/67850869-8bc91780-fb09-11e9-8339-aed1400a7835.png) 

5. Run segment rebuild
6. Check sql query in  logs on your dev enviroment
7. If today is 30. 10. 2019, you should see

`SELECT count(leadIdPrimary) count, max(leadIdPrimary) maxId, min(leadIdPrimary) minId FROM (SELECT DISTINCT orp.lead_id as leadIdPrimary, orp.lead_id as id, orp.leadlist_id FROM lead_lists_leads orp LEFT JOIN (SELECT l.id FROM leads l WHERE ((l.custom_date LIKE '2019-10-31%') AND (l.custom_date_time LIKE '2019-11-01%')) OR ((l.custom_date >= '2019-10-31') AND (l.custom_date_time >= '2019-11-01'))) members ON members.id=orp.lead_id WHERE (orp.leadlist_id = 9) AND (members.id IS NULL) AND (orp.manually_added = '0')) sss`

What is wrong. Compare for `date` should be  2019-11-01 and for `datetime` should be for equal  '2019-11-01%' and for gt/gte/lt/lte  should UTC date with time frame 

#### Steps to test this PR:
1. Apply this PR
2. Run segment:rebuild 
3. You should see now in debug on dev enviroment

`SELECT count(leadIdPrimary) count, max(leadIdPrimary) maxId, min(leadIdPrimary) minId FROM (SELECT DISTINCT orp.lead_id as leadIdPrimary, orp.lead_id as id, orp.leadlist_id FROM lead_lists_leads orp LEFT JOIN (SELECT l.id FROM leads l WHERE ((l.custom_date LIKE '2019-11-01%') AND (l.custom_date_time LIKE '2019-11-01%')) OR ((l.custom_date >= '2019-11-01') AND (l.custom_date_time >= '2019-11-01 11:48:12'))) members ON members.id=orp.lead_id WHERE (orp.leadlist_id = 9) AND (members.id IS NULL) AND (orp.manually_added = '0')) sss`

- l.custom_date LIKE '2019-11-01%' - ✔️ 
- l.custom_date_time LIKE '2019-11-01%' - ✔️ 
- l.custom_date >= '2019-11-01' - ✔️ 
- l.custom_date_time >= '2019-11-01 11:48:12' ✔️ 

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9784"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
